llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp

```diff
@@ -2232,8 +2366,11 @@ DSEState::eliminateDeadDefs(const MemoryLocationWrapper &KillingLocWrapper) {
     }
     MemoryDefWrapper DeadDefWrapper(
         cast<MemoryDef>(DeadAccess),
-        getLocForInst(cast<MemoryDef>(DeadAccess)->getMemoryInst()));
-    MemoryLocationWrapper &DeadLocWrapper = *DeadDefWrapper.DefinedLocation;
+        getLocForInst(cast<MemoryDef>(DeadAccess)->getMemoryInst(),
+                      /*ConsiderInitializesAttr=*/false));
+    assert(DeadDefWrapper.DefinedLocations.size() == 1);
+    MemoryLocationWrapper &DeadLocWrapper =
+        DeadDefWrapper.DefinedLocations.front();
```

Comment:
I'm not 100% sure what's going on here, but it seems weird that we have two modes for `MemoryDefWrapper`, a single MemoryLocation version here and a multiple MemoryLocation below. is there any way to make this a little less hacked together? why does this have to be a single MemoryLocation?

Reply:
We cannot apply the `initializes` attribute to DeadAccess/DeadDef since it would consider a call instruction as dead store and remove it incorrectly. Added a comment to explain it. Any suggestions?

Reply:
after staring at this a bit, I believe this preexisting code is conflating two things: it's assuming that if there is a memory location that `DeadDefWrapper` writes to that overlaps with `KillingLocWrapper`, it must have no other side effects and be deletable. this happens to be true for stores and libcalls like `strcpy` that are handled here, but is not necessarily true in general.

I think ideally we change `isRemovable` to be more accurate about arbitrary function calls, and check that here, but I'm ok with a TODO saying something like `TODO: this conflates the existence of a MemoryLocation with being able to delete the instruction. fix isRemovable() to consider calls with side effects that cannot be removed, e.g. calls with the initializes attribute, and remove getLocForInst(ConsiderInitializesAttr = false) workaround

Reply:
this still needs a comment

Reply:
Ooh, missed this comment. Added a TODO about isRemovable(). Thanks!